File: src/adapters/state/fileStateStore.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { z } from "zod";
import { AppState } from "../../domain/model.js";
import { StateStore } from "../../ports/stateStore.js";

const schema = z.object({
  lastRunJstDate: z.string().optional(),
  itemsState: z.record(
    z.object({
      score: z.number(),
      comments: z.number(),
      lastSeen: z.string()
    })
  ),
  postedHashes: z.record(z.string())
});

export class FileStateStore implements StateStore {
  constructor(private readonly filePath: string) {}

  async load(): Promise<AppState> {
    try {
      const body = await readFile(this.filePath, "utf8");
      return schema.parse(JSON.parse(body));
    } catch {
      return {
        itemsState: {},
        postedHashes: {}
      };
    }
  }

  async save(state: AppState): Promise<void> {
    await mkdir(path.dirname(this.filePath), { recursive: true });
    await writeFile(this.filePath, JSON.stringify(state, null, 2), "utf8");
  }
}

File: internal/adapters/state/file_state_store.go
package state

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"

	"github.com/HiroVodka/trendog/internal/domain"
)

type FileStateStore struct{ Path string }

func (s FileStateStore) Load(_ context.Context) (domain.AppState, error) {
	b, err := os.ReadFile(s.Path)
	if err != nil {
		return domain.AppState{ItemsState: map[string]domain.ItemState{}, PostedHashes: map[string]string{}}, nil
	}
	var st domain.AppState
	if err := json.Unmarshal(b, &st); err != nil {
		return domain.AppState{ItemsState: map[string]domain.ItemState{}, PostedHashes: map[string]string{}}, nil
	}
	if st.ItemsState == nil {
		st.ItemsState = map[string]domain.ItemState{}
	}
	if st.PostedHashes == nil {
		st.PostedHashes = map[string]string{}
	}
	return st, nil
}

func (s FileStateStore) Save(_ context.Context, st domain.AppState) error {
	if st.ItemsState == nil {
		st.ItemsState = map[string]domain.ItemState{}
	}
	if st.PostedHashes == nil {
		st.PostedHashes = map[string]string{}
	}
	if err := os.MkdirAll(filepath.Dir(s.Path), 0o755); err != nil {
		return err
	}
	b, _ := json.MarshalIndent(st, "", "  ")
	return os.WriteFile(s.Path, b, 0o644)
}
